main.go

```diff
@@ -1,117 +1,142 @@
-	"github.com/giantswarm/cluster-api-provider-kvm/service"
+	"flag"
+	"os"
+
+	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
+	// to ensure that exec-entrypoint and run can make use of them.
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/klog/klogr"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/util/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+
+	"github.com/giantswarm/cluster-api-provider-kvm/controllers"
+	//+kubebuilder:scaffold:imports
 )
 
 var (
-	f *flag.Flag = flag.New()
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
 )
 
-func main() {
-	err := mainE(context.Background())
-	if err != nil {
-		panic(microerror.JSON(err))
-	}
+func init() {
+	klog.InitFlags(nil)
+
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+
+	//+kubebuilder:scaffold:scheme
 }
 
-func mainE(ctx context.Context) error {
-	var err error
+var (
+	metricsAddr          string
+	enableLeaderElection bool
+	probeAddr            string
+)
 
-	var logger micrologger.Logger
-	{
-		c := micrologger.Config{}
+func main() {
 
-		logger, err = micrologger.New(c)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-	}
+	initFlags(pflag.CommandLine)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+	pflag.Parse()
 
-	// We define a server factory to create the custom server once all command
-	// line flags are parsed and all microservice configuration is storted out.
-	serverFactory := func(v *viper.Viper) microserver.Server {
-		// Create a new custom service which implements business logic.
-		var newService *service.Service
-		{
-			c := service.Config{
-				Logger: logger,
-
-				Flag:  f,
-				Viper: v,
-			}
-
-			newService, err = service.New(c)
-			if err != nil {
-				panic(microerror.JSON(err))
-			}
-
-			go newService.Boot(ctx)
-		}
-
-		// Create a new custom server which bundles our endpoints.
-		var newServer microserver.Server
-		{
-			c := server.Config{
-				Logger:  logger,
-				Service: newService,
-
-				Viper: v,
-			}
-
-			newServer, err = server.New(c)
-			if err != nil {
-				panic(microerror.JSON(err))
-			}
-		}
-
-		return newServer
-	}
+	ctrl.SetLogger(klogr.New())
 
-	// Create a new microkit command which manages our custom microservice.
-	var newCommand command.Command
-	{
-		c := command.Config{
-			Logger:        logger,
-			ServerFactory: serverFactory,
-
-			Description: project.Description(),
-			GitCommit:   project.GitSHA(),
-			Name:        project.Name(),
-			Source:      project.Source(),
-			Version:     project.Version(),
-		}
-
-		newCommand, err = command.New(c)
-		if err != nil {
-			return microerror.Mask(err)
-		}
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
```

Comment:
I'm slightly concerned about `ctrl.GetConfigOrDie()` - it creates the rest config used to communicate with kubernetes ([link](https://github.com/kubernetes-sigs/controller-runtime/blob/8b55f85c90c3b1df1af58ddb7fe50096fdc7aa99/pkg/client/config/config.go#L43-L61)). In kvm-operator we currently only [pass](https://github.com/giantswarm/kvm-operator/blob/29ac67e2def2f98bbb821eedbc53e37428e69b34/service/service.go#L68-L83) the address (from default flag value) and this flag was deprecated and then removed from upstream. It somehow works for `aws` and `azure` though so it might be ok.

Reply:
Seems fine for now. This is easy enough to adjust or write our own replacement.